Add Huum integration

frwickst opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add the Huum sauna integration via the Huum API.

There was a [previous try](https://github.com/home-assistant/core/pull/72674) at this PR, but it was abandoned due to a lack of time
this is a new try, with more time :) This PR implements fixes for the comments
in that PR as well.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30505

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

frwickst committed:
Add Huum integration

fireheadman commented:
Let us know when you need testers....

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    }
+)
+
+
```
This function is fairly small, I would inline it in where you use it

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        raise InvalidAuth from client_error
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class HuumConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```

frwickst replied:
Fixed

joostlek commented on homeassistant/components/huumtest/__init__.py:
```diff
@@ -0,0 +1,34 @@
+"""The A test integration for Huum Saunas integration."""
```
Please remove all the huumtest files

frwickst replied:
Ah, forgot to push the change, sorry, will fix

joostlek commented on tests/components/huum/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+    """Test we get the form."""
+    TEST_USERNAME = "test-username"
+
+    await setup.async_setup_component(hass, "persistent_notification", {})
```
This is not needed?

frwickst replied:
Could very well be yes, I copied some of the tests from and modified. There are just so many things to keep in mind implementing everything 😄  I'll remove.

frwickst replied:
Remove from the code, tests seem to do fine without it as well so all good.

joostlek commented on tests/components/huum/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": TEST_USERNAME,
+                "password": "test-password",
```
Constants

joostlek commented on tests/components/huum/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        )
+
+    assert result2["type"] == RESULT_TYPE_FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
```
Please finish config flows in either Create entry or abort so we also test that the config flow can recover from an error

frwickst replied:
Sorry, but I don't fully grasp what you are after here. Could you please point me in the right direction?

joostlek replied:
Well, after testing that you are redirected back to the form, we should test that the form is still able to complete. So what you do is you patch the lib again to succeed and submit again

frwickst replied:
Aaah, alright, sorry, then I understand what you mean 👍  I rather ask stupid questions here than change the code over and over, better to get it at least close to the right thing on the first try 😄 Thank you once again 🙇

joostlek replied:
No problem :)

frwickst replied:
Added to the tests 👍

joostlek commented on tests/components/huum/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.huum.config_flow.Huum.status",
+        side_effect=CannotConnect,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == RESULT_TYPE_FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_unknown_error(hass: HomeAssistant) -> None:
```
Checkout `@pytest.mark.parametrize`. you can make all these error tests parameterized so you don't have to have the same test logic thrice

frwickst replied:
I agree that there are some redundant code here, but I personally like to keep tests separate like this to have better naming and better control over possible extra data that might be needed when possible future issues needs to be tested with regards to the same errors. Personal opinion I guess, hope that you are OK with me leaving it without the parameterization, but I can change if that is something that is required by the codebase.

joostlek replied:
I mean, I am really a fan of keeping tests small and maintainable. With the requirement that config entry tests should end in create entry or abort, your tests will get big and parametrize is a perfect tool for this.

Keep in mind, design code on what you think is going to change. The only changes I usually see with config flows are 2 things: 1. a big overhaul where everything is changed 2. an extra exception (maybe your sauna needs to be on when you set it up to collect all data, and you found a way to catch that error in the lib and want to raise a separate error to give the user a better error message).

With the first case you probably always need to rewrite the tests, with the second one you just have to add an extra testcase and you are done.

This isn't a blocker, but I do have a strong preference

frwickst commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,135 @@
+        if self.hvac_mode == HVACMode.HEAT:
+            await self._turn_on(temperature)
+
+    async def async_update(self) -> None:
+        """Get the latest status data.
+
+        We get the latest status first from the status endpoints of the sauna.
+        If that data does not include the temperature, that means that the sauna
+        is off, we then call the off command which will in turn return the temperature.
+        This is a workaround for getting the temperature as the Huum API does not
+        return the target temperature of a sauna that is off, even if it can have
+        a target temperature at that time.
+        """
+        self._status = await self._huum_handler.status_from_status_or_stop()
```
Huum does not provide other things that I know of at the moment, so I think we are good

joostlek commented:
Feel free to resolve comments if you think you solved them and all questions are answered :)

joostlek commented:
Oh and please checkout the CI, stuff is failing

frwickst commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,135 @@
+        self._huum_handler = huum_handler
+
+    @property
+    def name(self) -> str:
```
By "adding a DeviceInfo" do you refer to this documentation https://developers.home-assistant.io/docs/device_registry_index?_highlight=deviceinfo#defining-devices

So I would add a `device_info` property function that returns a `DeviceInfo`?

joostlek replied:
or `_attr_device_info` in the constructor, which might be cleaner to do

joostlek replied:
But yes

frwickst replied:
In this case, I can't get a unique identifier of the sauna entity. Is it OK to just hard-code the name "Huum Sauna" as the name, so that the settings would be something in the line of:
```python
    self._attr_device_info = DeviceInfo(
        name="Huum Sauna",
        manufacturer="Huum",
    )
```

joostlek replied:
Please then set the config_entry.entry_id as unique_id for both the entity and the device

frwickst replied:
Sorry, this is my first integration, and there is quite a bit to learn here 😄 

How do I go about doing that? 
Is something like this what you are looking for?
``` python
class HuumDevice(ClimateEntity):
    self._attr_device_info = DeviceInfo(
        name="Huum Sauna",
        manufacturer="Huum",
    )
    def __init__(self, status: HuumStatusResponse, huum_handler: Huum) -> None:
        self.unique_id  = self.entity_id
```

joostlek replied:
1. Add `unique_id` as parameter to HuumDevice
2. in `async_setup_entry` fill it with `config_entry.entry_id`.


```py
def __init__(self, huum_handler: Huum, unique_id: str) -> None:
    self._attr_unique_id = unique_id
    self._attr_device_info = DeviceInfo(
        identifiers={(DOMAIN, unique_id)},
        name="Huum sauna",
    )
```

frwickst replied:
Thank you for all the help. Changed the code accordingly now.

frwickst commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,135 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Huum sauna with config flow."""
+    huum_handler = Huum(
+        entry.data[CONF_USERNAME],
+        entry.data[CONF_PASSWORD],
+        session=async_get_clientsession(hass),
+    )
+
+    try:
+        entry_status = await huum_handler.status_from_status_or_stop()
+    except Exception as status_error:
+        raise PlatformNotReady(status_error) from status_error
```
Just to be clear, would this be what you are looking for ?
```python
async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the Huum sauna with config flow."""
    huum_handler = hass.data.setdefault(DOMAIN, {})[entry.entry_id]

    try:
        entry_status = await huum_handler.status_from_status_or_stop()
    except Exception as status_error:
        raise PlatformNotReady(status_error) from status_error

    async_add_entities([HuumDevice(entry_status, huum_handler)])
```

joostlek replied:
```py
async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the Huum sauna with config flow."""
    huum_handler = hass.data.setdefault(DOMAIN, {})[entry.entry_id]

    async_add_entities([HuumDevice(huum_handler)], True)
```

frwickst replied:
So I don't need to get the status of the divide when settings it up?
Well, I'm all for less code, so I can change it to the way to suggested 😄

joostlek replied:
The explanation is somewhere down in one of the hidden comments (github collapsed them because there were too many)

frwickst replied:
Fixed according to suggestion

frwickst commented on homeassistant/components/huum/__init__.py:
```diff
@@ -0,0 +1,46 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Huum from a config entry."""
+    # hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
Fixed

joostlek commented:
Ah, we just encountered another issue. Pydantic.

Your lib is using pydantic v2. Which is incompatible with pydantic v1. I am trying to get people to make their libs compatible with pydantic v2 so we can migrate to it at some point. In issue #99218 you can see that we are not there yet and that it will still probably take a while.

So for this PR, there are 2 options:
1. Wait for pydantic v2 (which can be anything from a few months up to a year or so)
2. Downgrade to pydantic v1 (and add support for pydantic v2)
3. Use a completely different lib for whatever you use pydantic for. (I used it for mapping stuff from an API to python objects, there are other libs for that like Mashumaro and plain old dataclasses)

This is a blocker.

frwickst commented:
@joostlek 
Ah, no problem at all, I can downgrade to Pydantic 1 again, only upgraded to that today when I updated the underlying python package.

joostlek commented:
Please checkout some PRs in the issue I linked. Pydantic v2 has v1 packaged inside of it, so there's a trick to be able to support both (and this is a hard requirement for me, otherwise we have another new library that is blocking the upgrade)

joostlek commented:
Please also take in account that there are hidden comments. And could you maybe not force push, now its harder to see what changed in between of commits

frwickst commented:
I can skip force pushing, no problem.

frwickst committed:
Use DeviceInfo instead of name property for huum climate

frwickst committed:
Simplify entry setup for huum climate entry

frwickst committed:
Don’t take status as attribute for huum climate init

frwickst committed:
Remove unused import

frwickst committed:
Set unique id as entity id in huum init

frwickst committed:
Remove unused import for huum climate

frwickst committed:
Use entry ID as unique ID for device entity

frwickst committed:
Remove extra newline in huum climate

frwickst committed:
Upgrade pyhuum to 0.7.4

This version no longer users Pydantic

frwickst commented:
@joostlek I migrated away from Pydantic in the `pyhuum` package, so that is now taken care of 😄

frwickst committed:
Parameterize error huum tests

frwickst committed:
Update all requirements after pyhuum upgrade

joostlek commented:
If its ready for review, please press the button. Also reminder that there are still 3-4 open comments hidden

frwickst commented on homeassistant/components/huum/__init__.py:
```diff
@@ -0,0 +1,46 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Huum from a config entry."""
+    # hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    hass.data.setdefault(DOMAIN, {})
```
Fixed

frwickst committed:
Use Huum specific naming for ConfigFlow

frwickst committed:
Use constants for username and password in huum config flow

frwickst committed:
Use constants for temperature units

frwickst committed:
Fix typing and pylint issues

frwickst committed:
Update pyhuum to 0.7.5

frwickst committed:
Use correct enums for data entry flow in Huum tests

frwickst committed:
Remove test for non-thrown CannotConnect in huum flow tests

frwickst committed:
Refactor failure config test to also test a successful flow after failure

frwickst commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required("username"): str,
+        vol.Required("password"): str,
```
Fixed (along with some other constants in tests that was deprecated)

frwickst commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,135 @@
+    except Exception as status_error:
+        raise PlatformNotReady(status_error) from status_error
+
+    async_add_entities([HuumDevice(entry_status, huum_handler)])
```
Fixed

frwickst commented:
Waiting for all tests to pass and then I'll mark it as "Ready for review"

Thank you for all your help @joostlek ! All your comments should now be addressed (and I hope they are correct). Your swift feedback encouraged me to fix stuff right away! This was the issue last time where review times took so much time and the time between answers was either infinite (no answer) or took a bit too long so my attention span ran out 😄 

One more question. The "docs-missing" CI job is still failing, is that expected until the docs has been merged or is that something that I should fix in some way?

joostlek commented:
Add a link to the docs in the description :)

frwickst committed:
Fix ruff-format issues

frwickst commented:
Of course 😄 too easy so I missed it, thank you!

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,127 @@
+        )
+
+        self._huum_handler = huum_handler
+        self._status = None
```
This one can be set outside of the constructor. Can we type it?

frwickst replied:
Move from constructor to a class variable and typed.

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,127 @@
+        """
+        self._status = await self._huum_handler.status_from_status_or_stop()
+
+    async def _turn_on(self, temperature) -> None:
```
Can we type this?

frwickst replied:
Typing added

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,127 @@
+    @property
+    def current_temperature(self) -> float | None:
+        """Return the current temperature."""
+        return getattr(self._status, "temperature", None) or self._attr_min_temp
```
What is the third attribute of the `getattr`? If `self._status.temperature` does not exist, (without the third attribute) what does it return?

Also, otherwise we return the minimum temperature, but if the device can't report a current temperature, wouldn't it be better to return None as we don't know the current temperature?

frwickst replied:
The third value is the default value, there is no point in ever using `getattr` without a third argument.

So if `self._status.temperature` is not set, the `getattr` call will return `None` which in turn will trigger the `or` and return `self._attr_min_temp`.

And we can return `None` as well yes, then I'll just remove the `or` statement

joostlek replied:
But what i mean is,

```py
getattr(self._status, "temperature") == getattr(self._status, "temperature", None)
```

Is this true or false

joostlek replied:
Just looked it up, that's not true, left would raise AttributeError.

Also wondering, if we can type `self._status` can we just return `self._status.temperature` directly?

frwickst replied:
Exactly, it will raise `AttributeError`. Using `getattr` without a default value will is exactly the same as not using it at all.

So doing `self._status.temperature` is exactly the same as doing `getattr(self._status, "temperature")`, using `getattr` will do nothing here. There is actually a Ruff check for this as well, but I can't remember the plugin that needs to be enabled for it.

frwickst replied:
Typing `_status` will not set any default values, it will just set the type of the variable, so I don't see how that would help.

The type of `_status` is `_status: HuumStatusResponse | None = None`, only after `async_update` has been called can we be sure that `_status` is not `None`, meaning that we have to type it as `None` as well since `async_update` can't be called before the constructor has been called.

joostlek replied:
I mean that sounds good to me, just have a None check in the places where you return a value

frwickst replied:
Refactored

frwickst replied:
Here is the `ruff` rule that I mentioned earlier btw 🙂  Might be worth enabling for the entire codebase at some point.

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,127 @@
+    @property
+    def target_temperature(self) -> int:
+        """Return the temperature we try to reach."""
+        return (
+            getattr(self._status, "target_temperature", None)
+            or self._target_temperature
+        )
```
Same questions

frwickst replied:
If we return `None` from this then `async_set_hvac_mode` will need some refactoring that checks if `target_temperature` is None. But I can do it that way as well, it will not result in less code, but it will make this method smaller.

joostlek replied:
I think these changes are nice to make the code more robust and true to the actual state of the sauna.

frwickst replied:
Refactored

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
```
This function is small, can we inline it? This way you can omit the InvalidAuth error

frwickst replied:
Refactored

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        )
+
+
+class CannotConnect(HomeAssistantError):
```
Unused

frwickst replied:
Removed

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception as err:  # pylint: disable=broad-except
+            _LOGGER.error(str(err))
```
```suggestion
            _LOGGER.exception("Unknown error")
```

frwickst replied:
Would this not throw away information about what went wrong? So if some exception that I have not forseen is thrown, we would silence it and just log "Unknown error" instead of outputting the actual error in the logs. Just outputting a string like this would make it hard to debug no?

joostlek replied:
that's the funny thing. `LOGGER.exception` actually logs the whole exception. I also did not know this was a thing but its beautiful

frwickst replied:
Refactored

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+            await validate_input(self.hass, user_input)
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception as err:  # pylint: disable=broad-except
```
```suggestion
        except Exception:  # pylint: disable=broad-except
```

frwickst replied:
I use the `err` to log on the next line so removing it would not work.

frwickst replied:
Fixed

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
```
Consider inverting the config flow, this way you only have to have one `self.async_show_form`

frwickst replied:
Refactored

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        raise InvalidAuth from client_error
+
+
+class HuumConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
We aren't checking if the integration was already setup with the same data. Can someone has 2 saunas on one account? Can we get some kind of unique identifier?

frwickst replied:
As far as I can tell, you only have a single sauna connected to an account, which is strange. The API does not return any unique identifiers that I can tell. So there are no unique ids for the sauna itself. So the unique identifier would then be the email used for authenticating I guess.

joostlek replied:
> As far as I can tell, you only have a single sauna connected to an account, which is strange

when you check the data you get, do you get a list back or just an object?

I think it would be nice if we can set something unique and then also validate if this has been added before (there are quick function for that, this all can be done in 2 lines, so if you have more lines you are overcomplicating 😇 )

joostlek replied:
Oh and when we do this, please create a test for it

frwickst replied:
I get a single object back. So no list, no multiples, just a single entity. You can see the code in the `pyhuum` package here that calls the `/status` endpoint. https://github.com/frwickst/pyhuum/blob/main/huum/huum.py#L119

During auth there is only username and password that is passed as well, no identifier or extra value either. This would all point to there only being a single sauna that is linked to one account.

frwickst replied:
The best possible unique identifier that I can think of in this case, is the email of the user.

joostlek replied:
Can someone change their emailadress at huum?

frwickst replied:
Alright, so I started to dig a bit deeper here since I also found it a bit strange that there was no identifier for the sauna. And yes, there is an identifier in the app for the sauna, but it is not accessible through the API specs that Huum provided me. So I will have to do some MiTM proxying in order to get what kind of calls that they do. This of course brings up the question regarding if this is a "good thing to do" as Huum did not provide these endpoints to me. And HA calling them could lead to thing such as them changing the APIs in the future or in the worst case blocking access for some reason. But I will at least try to get the identifier now, I'll get back to you when I've tested things.

Regarding email, it does look like the email can be changed, I've not tried it however and doing so will break the auth since it is using basic auth as their way of authentication in the API.

frwickst replied:
So this still leads to the end result of there just being a single sauna linked to the account, but we might be able to get a unique identifier from Huum for said sauna.

joostlek replied:
Preferably we have an identifier (so if someone resetups the integration not all automations breaks down). Otherwise we just have to check on mail address.

frwickst replied:
Alright, so I MiTM proxied the app calls, and yes, there are things there are are used by the app to get more data, such as the sauna id and there also seem to be an endpoint to actually set the target temperature without turning on the sauna. The issue is that all of these endpoints are _not_ under the same path as the API docs that were provided by Huum. They also require a session token at times to get the value of the profile for instance, so the entire auth system is different for getting the profile data for instance.

So now the question becomes, are these things "OK" to call from Huums side. I would personally feel better getting an OK from them before starting to use them.

joostlek replied:
If the whole auth system is different then I think it's good to exclude it for now. I think checking if the current data already exists would be enough in the config flow

frwickst replied:
Alright, then I will look into having a unique identifier using email later on today I would say. I have emailed Huum now also and asked if they could have the profile API to be "public" as well, in order for me to use it. They are however on Christmas leave until the 1st of January, meaning that it will take some time for them to respond. We can also implement that logic later I would hope 🙂

joostlek replied:
I mean we can do that part later. We have stuff in HA that can make migrations like these easy :)

frwickst replied:
Ah, and it needs `self._abort_if_unique_id_configured()` as well of course to abort the setup process 😄

joostlek replied:
Yes and no, now we know there might be a really unique id we can set later on, I'd prefer to set that as unique id when we can. For now you can use `self._async_abort_entries_match({CONF_USERNAME: user_input[CONF_USERNAME]})`

This will check the entry data of other entries if the username is taken already

frwickst replied:
Hmmm.. doing this does not seem to stop the creation of a new entity, do I need to change `async_set_unique_id` in some other way as well. Setting it to `{CONF_USERNAME: user_input[CONF_USERNAME]}` did not seem to do it at least.

frwickst replied:
Nevermind me, this was a test issue, just needed to pass some extra data to the mock :)

frwickst replied:
This should now be done, hopefully 😄

joostlek commented on tests/components/huum/test_config_flow.py:
```diff
@@ -0,0 +1,97 @@
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": TEST_USERNAME,
+                "password": TEST_PASSWORD,
```
```suggestion
                CONF_USERNAME: TEST_USERNAME,
                CONF_PASSWORD: TEST_PASSWORD,
```

frwickst replied:
Fixed

frwickst committed:
Move _status outside of __init__ and type it

frwickst committed:
Type temperature argument for _turn_on in huum climate

frwickst committed:
Use constants for auth in huum config flow test

frwickst committed:
Refactor validate_into into a inline call in huum config flow

frwickst committed:
Refactor current and target temperature to be able to return None values

frwickst committed:
Remove unused huum exceptions

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,128 @@
+    @property
+    def current_temperature(self) -> int | None:
+        """Return the current temperature."""
+        return getattr(self._status, "temperature", None)
```
```suggestion
        if (status := self._status) is not None:
            return status.temperature
        return None
```

frwickst replied:
What does this do that my code did not do?

frwickst replied:
I have to say that this feels a bit like going into the realm of bikeshedding now, but I'll change it according to your comment just to have the code in the way the HA staff sees it appropriate.

frwickst replied:
Refactored

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,128 @@
+
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set hvac mode."""
+        if not self.target_temperature:
+            raise HomeAssistantError("No target temperature set, can't turn on sauna")
```
Right so let me get this straight. What actions can you execute on the sauna and what will change? If I understand correctly, when the sauna is idle, everything is None. Can you set temperature without setting hvac mode? can you set hvac mode without setting temperature?

I think recently we also introduced the generic `turn_on` and `turn_off` functions for climate. Could they be useful?

frwickst replied:
> Can you set temperature without setting hvac mode?
This is not possible unfortunately

> Can you set hvac mode without setting temperature?
Also no, they are connected, you set the temperature at the same time as turning it on

joostlek replied:
Should they be sent combined to the device? Or can they be separated and only when both are sent it will turn on?

frwickst replied:
While I have not looked in to `turn_on`  and `turn_off` I don't really see them as being helpful here. The mode `HEAT` is kinda what you do and you turn on the sauna, and it makes the UX niced for the user as well when the UI looks like this
<img width="490" alt="image" src="https://github.com/home-assistant/core/assets/3242350/619e04e6-92c1-4c7a-9ba7-9ef86e0c322c">

whereas "on" and "off" would not really signal the same thing. But I might be wrong here and the on/off might do more than I know about.

joostlek replied:
Hmm, so in that light it makes sense to set a temperature directly when you select a HVAC mode.

frwickst replied:
>Should they be sent combined to the device? Or can they be separated and only when both are sent it will turn on?

They are sent combined yes, the "on" command also sets the temperature at the same time. There is no way (at least with the API docs I know of) to set the temperature separately from turning on the sauna. So turning it on and settings the temp is the same action in their API.

frwickst replied:
> Hmm, so in that light it makes sense to set a temperature directly when you select a HVAC mode.

By this, do you want me to refactor things back so that it does try to set the min temperature if no temperature is set, or should we just rely on the status to have been fetched by the time a user tried to turn on the sauna. The user would need to be mighty fast in order to turn on the sauna in between settings up the integration and the first `async_update` call I would say.

joostlek replied:
Honestly, I think you know better what the use case would be. I think its fine to turn it on on the minimum temperature.

When I responded that it should be removed was because I did not know you needed both to actually work so sorry for that.

frwickst replied:
This should now be resolved, I'm using the minimum temp if there is no target temp set. This should only be the case pretty much if someone has never used the sauna ever, and sets it up in home assistant and turns it on there before they try to do so manually. So this should be a uber-corner case if I'm not missing something, but it is now taken into consideration 😄

frwickst committed:
Flip if-statment in async_step_user flow setup to simplify code

frwickst committed:
Change current and target temperature to be more future proof

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+                _LOGGER.error("Could not log in to Huum with given credentials")
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.error("Unknown error")
```
```suggestion
                _LOGGER.exception("Unknown error")
```

frwickst replied:
Fixed

frwickst committed:
Log exception instead of error

joostlek commented:
Oh I just noticed you migrated to mashumaro, awesome :)

frwickst commented:
Yeah, that was just easier overall, this way I don't need to worry about migrating between Pydantic versions nor removing the version-check logic when HA starts supporting Pydantic V2 😄

rytilahti commented on homeassistant/components/huum/__init__.py:
```diff
@@ -0,0 +1,46 @@
+        await huum.status()
+    except aiohttp.ClientError as err:
+        _LOGGER.error("Could not log in to Huum with given credentials")
+        raise ConfigEntryNotReady(
+            "Could not log in to Huum with given credentials"
+        ) from err
```
`ConfigEntryAuthFailed` is usually raised for authentication issues as, IIRC, it offers the user a way to start the reauth flow. `ConfigEntryNotReady` will just keep retrying forever while showing the error message on the integration page without actionable items.

joostlek replied:
This is a valid point, but I like to have reauth added in a later PR as it's not needed for the initial integration PR. But I did not mention this yet, thanks for reminding.

frwickst replied:
Is it enough to just raise `ConfigEntryAuthFailed` or does there need to be more logic in place for things to work? I mean, I can just check the status code of the response and if it is 401 and then throw that error.

joostlek replied:
That should go in a follow up PR, it should also include a config flow where the user can reenter their password and restore their entry again

frwickst replied:
I guess that would need to be checked every place where an API call is made, so when checking status, turning on/off etc as well right?

joostlek replied:
I mean you have that concentrated in 2 places so its not a big of a deal :)

frwickst replied:
But that is enough then, no other code is needed, just raise the exception, and done?

I guess I should start using custom exceptions coming from my library instead of "leaking" `aiohttp` exceptions as well. I will try to make those changes as well.

rytilahti replied:
Good point on keeping it simple for the initial PR 👍 

For exception handling, it's indeed better to raise library specific custom exceptions, but you can still use "raise .. from ex" to keep the original exception around.

frwickst replied:
I'll keep it simple for now and skip implementing re-auth logic, I will however implement the custom exceptions so that it is easier to implement the re-auth when the time comes

frwickst replied:
With that I see this matter resolved for now 😄

rytilahti reviewed (commented):
Just something very minor I noticed wrt. auth failures.

Also, if you wish, you could also add `repository = <url>` into the `tool.poetry` table inside your pyproject.toml, this will create a link on the package's pypi page making it easy to find.

frwickst committed:
Use custom pyhuum exceptions

frwickst commented:
@rytilahti I added `repository` and `readme` to pyproject.toml of the library now, thanks for the suggestion 👍

joostlek commented:
So to get things straight, there are 2 things needed for this PR:

1. Checking for uniqueness in the config flow. Either via some kind of identifier or just checking if the email was already used.
2. Making sure the sauna is able to turn on.

(Also, feel free to message me on discord to have a chat :) )

frwickst commented:
For #2 I can assure you that it turns on (with this code), it is almost at 80C already 😅 

But I will change the logic still there to make sure that we can turn it on even if the target temp is `None` and in those cases set it to the minimum temp configured (40C)

frwickst committed:
Add checks for duplicate entries

frwickst committed:
Use min temp if no target temp has been fetched yet when heating huum

frwickst committed:
Fix tests so that mock config entry also include username and password

frwickst committed:
Fix ruff styling issues

I don’t know why it keeps doing this. I run `ruff` locally, and then it does not complain, but CI must be doing something else here.

joostlek commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -54,7 +54,9 @@ async def async_step_user(
                 await self.async_set_unique_id(
                     user_input[CONF_USERNAME], raise_on_progress=False
                 )
```
```suggestion
```
You can remove this one as you're now aborting based on entry data. In the future we could use the unique id here

frwickst replied:
Alright, I'll remove this 🙂

frwickst replied:
Fixed

frwickst commented:
@joostlek I now hope that I've addressed all of your points 😄 Feel free to add more things if there is still something that you feel is off, but hopefully we are getting to the point that most things are ironed out 😄

frwickst committed:
Remove unneded setting of unique id

joostlek commented:
Yep. I think I will wait until after the beta cut since I don't want to put more pressure on the beta release today

frwickst commented:
That means that if there are no more things to fix, this would make it to the 2024.2 release at the earliest, right? I'm not a hurry, just want to set my expectations 😄

joostlek commented:
I still need to take another look at it but I think its near finished. First going to cook some food

frwickst committed:
Update requirements

joostlek commented on homeassistant/components/huum/climate.py:
```diff
@@ -0,0 +1,133 @@
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set hvac mode."""
+        if hvac_mode == HVACMode.HEAT:
+            temperature = (
+                self.target_temperature
+                if self.target_temperature
+                else int(self.min_temp)
+            )
```
So if I understand correctly, target_temperature is now only set when the device is not idle. But since it's not idle, it's heating, and when its already heating, you dont have a reason to set the hvac mode to heat as it already is.

Furthermore, please avoid using multi line inline if statements if they span more than 1 line.

I will discuss this with someone else to make sure this is a right choice

frwickst replied:
The only way to set the target temperature through the API currently is by turning it on, that is unfortunately true. This is not a choice by me, but by the API Huum has provided me. I hope that I get an OK from Huum to start using the same API as they are using in their app soon enough, as that _can_ set the temperature separately, but as of right now that is not possible.

For the code style here, do you have a code style guide somewhere that I could follow, for me multi-line if-else statement like these are commonplace. I would rather focus on the actual logic of the code than the style of it, so it would be greatly helpful if I could check some place for how you want code formatted. With that said, I'll change this to a "normal" if-else statement.

frwickst replied:
But there is actually a bug in there as well, so I will have to fix that. The target temperature that the user has set is not actually used, and it jumps back to the one that we got from the API. I did not use the`self._target_temperature` that I had already created but has mistakenly not used 😄

joostlek replied:
apart from the docs I don't really know if there is a style guide. I am also still learning the little quirks of the style we use

frwickst replied:
I now added one more commit that fixes the settings of the current temperature. Due to the amount of refactoring done, I forgot to change the logic accordingly, sorry. So now the target temperature works and it can be set properly in the UI.

The issue previously was that if the sauna was off and the target temp was set in the UI and then the sauna was turned to heat, then the target temp from the API was used for turning it on and not the temp set by the user in the UI. That should now be fixed. It also simplified some of the code so that was nice :) I would really hope that we get access to the proper APIs, as that would simplify this further and would not require us to do some internal setting of temperature the way it is now done.

frwickst replied:
I guess it works something like this now (note that the code is the real thing to check, this is just a flow chart)

```mermaid
flowchart LR
    load[Load integration] --> status[Get status]
    onUpdate[On update] --> status
    userSet[User sets temperature in UI] --> setTempUser[Set `_target_temperature` from UI]
    setTempUser --> isOn{Is sauna on?}
    isOn -- Yes --> callTemp
    status --> hasTempOrOn{Do we have \n`_target_temperature` \nor is the sauna on?}
    hasTempOrOn -- Yes --> setTempStatus[Set `_target_temperature` from status]
    callTemp[Get `target_temperature`] --> hasTargetTemp{Is `_target_temperature` set?}
    hasTargetTemp -- No --> returnMin[Return min temp]
    hasTargetTemp -- Yes --> returnTemp[Return `_target_temperature`]
    returnMin --> turnOn[Turn on sauna with temperature]
    returnTemp --> turnOn[Turn on sauna with temperature]
```

(PS. made it left-to-right to not take up a ton of space in the PR discussion, top-down is slightly more readable)

joostlek replied:
(what, I did not even know github had support for this. What did you use to create this?)

frwickst replied:
It is [Mermaid](https://mermaid.js.org/) :) It has been supported [for a while](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/), it is also supported on GitLab

frwickst committed:
Refactor temperature setting to support settings target temparature properly

frwickst commented:
Please don't leave me hanging now @joostlek 😄  We had such a good start on this one. I don't mind the work, but it would be good to know what I still need to do to get this one to the finish line. There is of course no hurry to get this into a release, but I don't want this to grow stale either 😅

joostlek commented:
Don't worry, currently there is a bigger focus on fixing bugs etc for the 2024.1 release. After tomorrow I can poke someone to give this a quick look

fireheadman commented:
@joostlek     How did the review turn out?

joostlek commented:
I'll ask again today

fireheadman commented:
@joostlek    we are very curious how this is going?   Any updates for us?
Thanks Ben

frwickst commented:
@fireheadman  As of now, we are waiting for a second opinion. @joostlek marked this as `second-opinion-wanted` a couple of days ago.

So the status is the same as it was about a month ago, we need someone else to look at this before it can be merged, and make possible changes according to the review of the next person reviewing the code. So just to set your expectations @fireheadman , this will most likely not be in the next release.

rytilahti commented on homeassistant/components/huum/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+                _LOGGER.exception("Unknown error")
+                errors["base"] = "unknown"
+            else:
+                self._async_abort_entries_match(
+                    {CONF_USERNAME: user_input[CONF_USERNAME]}
```
I suppose there is no better way to identify the sauna than the username? A single user can have only a single sauna, is that correct?

frwickst replied:
I will refer you to this answer https://github.com/home-assistant/core/pull/106420#discussion_r1437024134

TL;DR A better identifier exists, but we don't have "explicit permission" to use that API as it is not under the same API prefix, it uses a different auth method and there is no documentation other than my MiTM testing for the API 😄  The same goes for the sauna id, the user id and settings temperature without turning on the sauna.

rytilahti replied:
Okay, I think this is then good enough for the time being! 👍

rytilahti reviewed (approved):
Looks rather straightforward to me, so in my opinion we could merge this for the next release assuming the one user, one sauna holds.
